core/commonMain/src/kotlinx/serialization/encoding/ReorderingCompositeEncoder.kt

```diff
@@ -0,0 +1,240 @@
+        index: Int,
+        crossinline encoder: CompositeEncoder.() -> Unit
+    ) {
+        if (callShouldEncodeElementDefault) {
+            bufferedCalls[mapElementIndex(descriptor, index)] = BufferedCall(index) {
+                if (shouldEncodeElementDefault(descriptor, index)) {
+                    encoder()
+                }
+            }
+        } else {
+            bufferedCalls[mapElementIndex(descriptor, index)] = BufferedCall(index) {
+                encoder()
+            }
+        }
+        callShouldEncodeElementDefault = false
+    }
+
+    override fun endStructure(descriptor: SerialDescriptor) {
+        encodeBufferedFields(descriptor)
+    }
+
+    private fun encodeBufferedFields(descriptor: SerialDescriptor) {
+        val classEncoder = beginStructure(descriptor)
+        bufferedCalls.forEach { fieldToEncode ->
+            fieldToEncode?.encoder?.invoke(classEncoder)
+        }
+        classEncoder.endStructure(descriptor)
+    }
+
+    override fun encodeBooleanElement(descriptor: SerialDescriptor, index: Int, value: Boolean) {
+        bufferEncoding(descriptor, index) { encodeBooleanElement(descriptor, index, value) }
+    }
+
+    override fun encodeByteElement(descriptor: SerialDescriptor, index: Int, value: Byte) {
+        bufferEncoding(descriptor, index) { encodeByteElement(descriptor, index, value) }
+    }
+
+    override fun encodeCharElement(descriptor: SerialDescriptor, index: Int, value: Char) {
+        bufferEncoding(descriptor, index) { encodeCharElement(descriptor, index, value) }
+    }
+
+    override fun encodeDoubleElement(descriptor: SerialDescriptor, index: Int, value: Double) {
+        bufferEncoding(descriptor, index) { encodeDoubleElement(descriptor, index, value) }
+    }
+
+    override fun encodeFloatElement(descriptor: SerialDescriptor, index: Int, value: Float) {
+        bufferEncoding(descriptor, index) { encodeFloatElement(descriptor, index, value) }
+    }
+
+    override fun encodeIntElement(descriptor: SerialDescriptor, index: Int, value: Int) {
+        bufferEncoding(descriptor, index) { encodeIntElement(descriptor, index, value) }
+    }
+
+    override fun encodeLongElement(descriptor: SerialDescriptor, index: Int, value: Long) {
+        bufferEncoding(descriptor, index) { encodeLongElement(descriptor, index, value) }
+    }
+
+    override fun <T : Any> encodeNullableSerializableElement(
+        descriptor: SerialDescriptor,
+        index: Int,
+        serializer: SerializationStrategy<T>,
+        value: T?
+    ) {
+        bufferEncoding(descriptor, index) { encodeNullableSerializableElement(descriptor, index, serializer, value) }
+    }
+
+    override fun <T> encodeSerializableElement(
+        descriptor: SerialDescriptor,
+        index: Int,
+        serializer: SerializationStrategy<T>,
+        value: T
+    ) {
+        bufferEncoding(descriptor, index) { encodeSerializableElement(descriptor, index, serializer, value) }
```

Comment:
This is incorrect. You will need a wrapping encoder so `Encoder.beginStructure` will work correctly. To make that work you also need a wrapping `SerializationStrategy` that captures the delegate encoder (for the wrapping) unless you add a way to determine the encoder in another way.

Reply:
I don't understand why it may not work. The serializer is not going to be called during the buffering step. FYI I already tested with a data class with a value class in a value class containing a list, and another value class containing a data class, and beginStructure is called only inside the `encodeSerializableValue`

Reply:
Say we have a structure: `{ a: { b: "foo", c: 42}, d: true }` then property `a` is encoded by the outer serializer (o) invoking `encodeSerializableElement(descriptor, 0, innerSerializer, value.a)`. This is then implemented by getting an `Encoder` instance and calling `innerSerializer.serialize(encoder, value)`. This `InnerSerializer.serialize` function then calls `beginStructure` (for the inner struct) which returns another `CompositeEncoder`.

If you implement this by delegating `encodeSerializableElement` to the non-reordered encoder it gets to decide what `CompositeEncoder` is used for the inner struct. If you interject it is possible for the ordered composite encoder to override it.

Reply:
But as you said, it's calling `encodeSerializableElement` so it won't reach the `innerSerializer` during buffering, but only during the next step when encoding with the CompositeEncoder delegate.